### PR TITLE
fix webpack-serve args definition by using WebpackServe.Options

### PR DESCRIPTION
**Pull request by Jokcy**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://github.com/webpack-contrib/webpack-serve#serveoptions>>
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by armanio123** [comment]
@rynclark Could you please review this PR? For some reason the typescript-bot didn't tag you.

**Comment by rynclark** [comment]
webpack-serve is still on 1.0.2, 1.1 hasn’t been released yet. This doesn’t match up with the node API too. Can you explain your changes please? @Jokcy

**Comment by Jokcy** [comment]
@rynclark I set version to 1.1 because I'm not sure whether types version should follow pkg version.
But I'm sure this matches the node API on webpack-serve 1.0.2, you can run my test code to confirm it.

**Comment by Jokcy** [comment]
@rynclark and I also find 'compiler-error' and 'compiler-warning' event definition is not correct. 

Based on API doc on webpack-serve, these two events accept args structure like this:
```ts
on(event: 'compiler-error', listener: (args: {compiler: webpack.Compiler, json: any}))
```
the key name should be json, not stats, below is the output I got when `compiler-error`:
```
{ json:
   { errors:
      [ './app/client/components/Hello.jsx\nModule build failed: SyntaxError: /Users/Jokcy/workspace/personal/2018/my-oauth/app/client/components/Hello.jsx: Unexpected token, expected "..." (4:40)\n\n  2 | import { hot } from \'react-hot-loader\'\n  3 | \n> 4 | const Hello = (props) => <h1Hello from {props.compiler} and123 {props.framework}!</h1>\n    |                                         ^\n  5 | \n  6 | export default hot(module)(Hello)\n    at _class.raise (/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/@babel/parser/lib/index.js:3861:15)\n    at _class.unexpected (/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/@babel/parser/lib/index.js:5190:16)\n    at _class.expect (/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/@babel/parser/lib/index.js:5178:28)\n    at _class.jsxParseAttribute (/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/@babel/parser/lib/index.js:3437:14)\n    at _class.jsxParseOpeningElementAt (/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/@babel/parser/lib/index.js:3460:35)\n    at _class.jsxParseElementAt (/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/@babel/parser/lib/index.js:3484:33)\n    at _class.jsxParseElement (/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/@babel/parser/lib/index.js:3553:19)\n    at _class.parseExprAtom (/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/@babel/parser/lib/index.js:3560:21)\n    at _class.parseExprSubscripts (/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/@babel/parser/lib/index.js:5868:21)\n    at _class.parseMaybeUnary (/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/@babel/parser/lib/index.js:5847:21)\n @ ./app/client/index.jsx 7:36-65\n @ multi ./app/client/index' ],
     warnings: [],
     version: '4.10.0',
     hash: 'b4db816c44c0ca58fea6',
     time: 4868,
     builtAt: 1528340035782,
     publicPath: 'http://192.168.141.169:8888/public/',
     outputPath: '/Users/Jokcy/workspace/personal/2018/my-oauth/app/public',
     assetsByChunkName: { main: [Array] },
     assets: [ [Object], [Object], [Object] ],
     filteredAssets: 0,
     entrypoints: { main: [Object] },
     namedChunkGroups: { main: [Object] },
     chunks: [ [Object] ],
     modules:
      [ [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object] ],
     filteredModules: 0,
     children: [ [Object] ] },
  compiler:
   Compiler {
     _pluginCompat:
      SyncBailHook {
        _args: [Array],
        taps: [Array],
        interceptors: [],
        _call: [Function: lazyCompileHook],
        call: [Function: lazyCompileHook],
        _promise: [Function: lazyCompileHook],
        promise: [Function: lazyCompileHook],
        _callAsync: [Function: lazyCompileHook],
        callAsync: [Function: lazyCompileHook],
        _x: undefined },
     hooks:
      { shouldEmit: [SyncBailHook],
        done: [AsyncSeriesHook],
        additionalPass: [AsyncSeriesHook],
        beforeRun: [AsyncSeriesHook],
        run: [AsyncSeriesHook],
        emit: [AsyncSeriesHook],
        afterEmit: [AsyncSeriesHook],
        thisCompilation: [SyncHook],
        compilation: [SyncHook],
        normalModuleFactory: [SyncHook],
        contextModuleFactory: [SyncHook],
        beforeCompile: [AsyncSeriesHook],
        compile: [SyncHook],
        make: [AsyncParallelHook],
        afterCompile: [AsyncSeriesHook],
        watchRun: [AsyncSeriesHook],
        failed: [SyncHook],
        invalid: [SyncHook],
        watchClose: [SyncHook],
        environment: [SyncHook],
        afterEnvironment: [SyncHook],
        afterPlugins: [SyncHook],
        afterResolvers: [SyncHook],
        entryOption: [SyncBailHook] },
     name: undefined,
     parentCompilation: undefined,
     outputPath: '/Users/Jokcy/workspace/personal/2018/my-oauth/app/public',
     outputFileSystem: MemoryFileSystem { data: [Object] },
     inputFileSystem:
      CachedInputFileSystem {
        fileSystem: NodeJsInputFileSystem {},
        _statStorage: [Storage],
        _readdirStorage: [Storage],
        _readFileStorage: [Storage],
        _readJsonStorage: [Storage],
        _readlinkStorage: [Storage],
        _stat: [Function: bound bound ],
        _statSync: [Function: bound bound ],
        _readdir: [Function: bound readdir],
        _readdirSync: [Function: bound readdirSync],
        _readFile: [Function: bound bound readFile],
        _readFileSync: [Function: bound bound ],
        _readJson: [Function],
        _readJsonSync: [Function],
        _readlink: [Function: bound bound ],
        _readlinkSync: [Function: bound bound ] },
     recordsInputPath: undefined,
     recordsOutputPath: undefined,
     records:
      { 'html-webpack-plugin for "index.html"': [Array],
        modules: [Object],
        chunks: [Object],
        prepreHash: 'x',
        preHash: '08692b19caf8c628a71f',
        hash: 'b4db816c44c0ca58fea6',
        moduleHashs: [Object],
        chunkHashs: [Object],
        chunkModuleIds: [Object] },
     fileTimestamps: Map {},
     contextTimestamps: Map {},
     resolverFactory:
      ResolverFactory {
        _pluginCompat: [SyncBailHook],
        hooks: [Object],
        cache1: WeakMap {},
        cache2: [Map] },
     resolvers: { normal: [Object], loader: [Object], context: [Object] },
     options:
      { target: 'web',
        mode: 'development',
        devtool: 'source-map',
        entry: [Array],
        output: [Object],
        resolve: [Object],
        module: [Object],
        plugins: [Array],
        cache: true,
        context: '/Users/Jokcy/workspace/personal/2018/my-oauth',
        node: [Object],
        performance: false,
        optimization: [Object],
        resolveLoader: [Object] },
     context: '/Users/Jokcy/workspace/personal/2018/my-oauth',
     requestShortener:
      RequestShortener {
        currentDirectoryRegExp: /(^|!)\/Users\/Jokcy\/workspace\/personal\/2018\/my\-oauth/g,
        parentDirectoryRegExp: /(^|!)\/Users\/Jokcy\/workspace\/personal\/2018/g,
        buildinsAsModule: true,
        buildinsRegExp: /(^|!)\/Users\/Jokcy\/workspace\/personal\/2018\/my\-oauth\/node_modules\/webpack/g,
        cache: [Map] },
     running: true,
     watchFileSystem:
      NodeWatchFileSystem {
        inputFileSystem: [CachedInputFileSystem],
        watcherOptions: [Object],
        watcher: [EventEmitter] },
     dependencies: undefined,
     _lastCompilationFileDependencies:
      SortableSet {
        '/Users/Jokcy/workspace/personal/2018/my-oauth/.babelrc',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/app/client/components/Hello.jsx',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/app/client/index.jsx',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/lib/build/template.html',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/fbjs/lib/ExecutionEnvironment.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/fbjs/lib/camelize.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/fbjs/lib/camelizeStyleName.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/fbjs/lib/containsNode.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/fbjs/lib/emptyFunction.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/fbjs/lib/emptyObject.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/fbjs/lib/getActiveElement.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/fbjs/lib/hyphenate.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/fbjs/lib/hyphenateStyleName.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/fbjs/lib/invariant.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/fbjs/lib/isNode.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/fbjs/lib/isTextNode.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/fbjs/lib/shallowEqual.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/fbjs/lib/warning.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/lodash/lodash.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/loglevelnext/dist/loglevelnext.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/object-assign/index.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/prop-types/checkPropTypes.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/prop-types/lib/ReactPropTypesSecret.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/punycode/punycode.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/querystring-es3/decode.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/querystring-es3/encode.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/querystring-es3/index.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/react-dom/cjs/react-dom.development.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/react-dom/index.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/react/cjs/react.development.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/react/index.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/url/url.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/url/util.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/webpack-hot-client/client/hot.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/webpack-hot-client/client/index.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/webpack-hot-client/client/log.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/webpack-hot-client/client/socket.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/webpack/buildin/global.js',
        '/Users/Jokcy/workspace/personal/2018/my-oauth/node_modules/webpack/buildin/module.js',
        _sortFn: undefined,
        _lastActiveSortFn: undefined,
        _cache: undefined,
        _cacheOrderIndependent: undefined },
     _lastCompilationContextDependencies:
      SortableSet {
        _sortFn: undefined,
        _lastActiveSortFn: null,
        _cache: undefined,
        _cacheOrderIndependent: undefined } } }
```
I can merge this into one PR if needed

**Comment by Jokcy** [comment]
any feedback?

**Review by ryanclark — CHANGES_REQUESTED** [review]
See comment on PR

**Comment by Jokcy** [comment]
@rynclark Thx for your response. I reviewed the code of webpack-serve again and find you are right about `config` and `compiler` options. But I still find out the serve options will work either.

The serve function we get from `require('webpack-server')` use `getOptions(opts)` from `const getOptions = require('./lib/options')` which parse the options we pass into. In the `./lib/options` file we can see code:
```js
module.exports = (opts) => {
  const flags = opts.flags || {};
  const nodeVersion = parseInt(process.version.substring(1), 10);

  return load(opts).then((configs) => {
    const [first] = configs;
    const options = merge({}, defaults, opts, flags, first.serve);
    ....
  }
}
```
`load` do parse configs and compilers as what you said. The most important code is `const options = merge({}, defaults, opts, flags, first.serve);` which merge is `lodash/merge`. The options variable is merged from

**defaults**
the default options defined like this:
```js
const defaults = {
  clipboard: true,
  compiler: null,
  config: {},
  content: [],
  dev: { publicPath: '/' },
  flags: {},
  host: 'localhost',
  hot: {},
  http2: false,
  https: null,
  // https: {
  //   key: fs.readFileSync('...key'),
  //   cert: fs.readFileSync('...cert'),
  //   pfx: ...,
  //   passphrase: ...
  // },
  logLevel: 'info',
  logTime: false,
  open: false,
  // open: { app: <String>, path: <String> }
  port: 8080,
  protocol: 'http',
};
```

**opts**
the options we passed into when we invoke `webpack-serve`

**flags**
`opts.flags`, emmmmm... I didn't find out any doc yet. Looks like some options in flags will have the biggest weights.

**first.serve**
the serve options defined in webpack configuration

So basically I'm quite sure if we don't config serve options in webpack configuration, the 'serve options' we passed when we invoke the `webpack-serve` function will take effect.

I'm not sure if you will understand me clearly because of my poor English😂

**Comment by rynclark** [comment]
Ah yes, I missed the merge of `opts` and `first.serve`, I do apologise.

The only change needed for this PR is the version changed back to `1.0` (as it needs to reflect the version of `webpack-serve`), and optionally, the `stats` property changed to `json`.

Nice catch, please accept my apologies @Jokcy.

**Comment by Jokcy** [comment]
@rynclark Great! Now there are two problems left
1. `config` and `compiler` can accept array and single object, should I fix this together? Actually they should use single object as defined in doc😄
2. How to revision? just create another commit or should I using `--amend`?

**Review by ryanclark — APPROVED** [review]

**Comment by rynclark** [comment]
Sorry for the delay. Great job, @Jokcy

**Comment by rynclark** [comment]
Note for the maintainer: when installing the latest of
`@types/webpack-serve` (1.0.1) it still includes version 3.0.0 of
`@types/webpack-hot-client`, even though the latest version is 3.0.1. Any
ideas?

**Comment by sandersn** [comment]
@Jokcy thanks for your contribution!
@rynclark I just installed `@types/webpack-serve` (before merging this PR) and it is the same as the index.d.ts in master. Both say they are version 3.0. Is 3.0.1 the current version of the package itself maybe?

**Comment by rynclark** [comment]
It’s the dependency of `@types/webpack-hot-client` that gets installed alongside it that’s wrong - I’m getting 3.0.0, when the latest is 3.0.1. If I manually install `@types/webpack-hot-client@3.0.1`, it then puts 3.0.0 inside the `node_modules` folder inside `@types/webpack-serve`, so it doesn’t fix the issue. No idea why, as the dependencies are set to `*`, so it should grab the latest. @sandersn

**Comment by sandersn** [comment]
Oops, I meant to specify that I was talking about webpack-hot-client. In fact, I diffed them to make sure:

```sh
$ diff webpack-hot-client/index.d.ts  ~/src/test/node_modules/@types/webpack-hot-client/index.d.ts 
```

What version do you get if you just install `@types/webpack-hot-client`? Is it 3.0 or 3.0.1?

**Comment by rynclark** [comment]
My bad, it was my `yarn.lock` that fixed it to `3.0.0` after installing `@types/webpack-serve` originally (1.0.0), and yarn never updated that when I updated to 1.0.1. // @sandersn

**Comment by Jokcy** [comment]
Finally merged! I guess I can drop my typings soon😄
